hypothesis/frontend-shared
# Create new Popover component

> Depends on https://github.com/hypothesis/frontend-shared/pull/1789

Extract the popover element and the logic to dynamically position it from `Select` and into its own `Popover` module.

This will allow us to use this component in other contexts.

### Considerations

The popover currently requires an anchor element to be provided, and calculates its position and size based on that.

For now, I decided to keep some decisions made, like the fact that the popover appears below the anchor element unless there's not enough space, in which case it appears above it.

It also matches the anchor element's size, unless it can't fit its content, in which case it grows.

It should be possible to customize both of those (define initial position, determine how to calculate size, etc.), but I'm not going to change that as part of this PR.

We may also need to support a different way to determine the anchor, like providing a rectangle rather than an anchor element.

### Out of scope

1. I have finally decided not to add a documentation page for `Popover` in this PR, as it has enough changes already.
    Separately, I will add a documentation page explaining the component, how it works and behaves by default, and including some examples.
    EDIT: Addressed in https://github.com/hypothesis/frontend-shared/pull/1793
2. I will also make the Popover render its children only when it's open, making the behavior transparent for the Select and consistent anywhere where the Popover is used.
    EDIT: This is addressed in https://github.com/hypothesis/frontend-shared/pull/1792

### Testing steps

Sanity check that all examples in http://localhost:4001/input-select keep working as expected

### TODO

- [x] Add tests for popover
- [x] Remove some tests from `Select`, if they overlap with Popover tests
